File: 7-2-3.java
import java.util.*;

public class Main {
    static long solve(int u, List<Integer> A, List<List<Integer>> E) {
        long ret = A.get(u);
        for (int v : E.get(u)) {
            long ret2 = solve(v, A, E);
            if (ret2 > 0) ret += ret2;
        }
        return ret;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        List<List<Integer>> E = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            E.add(new ArrayList<>());
        }
        for (int i = 1; i < n; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            E.get(u).add(v);
        }
        List<Integer> A = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            A.add(sc.nextInt());
        }
        System.out.println(solve(0, A, E));
    }
}

File: 7-2-3.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10 ** 6)

def solution(n, A, edges):
    E = list([] for _ in range(n))
    for p, c in edges:
        E[p].append(c)

    return solve(0, A, E)

def solve(u, A, E):
    ret = A[u]
    for v in E[u]:
        ret2 = solve(v, A, E)
        if ret2 > 0:
            ret += ret2

    return ret

n = int(input())
edges = list(list(map(int, input().split())) for _ in range(n - 1))
A = list(map(int, input().split()))
print(solution(n, A, edges))